docs/rfcs/035-physical-replication.md

```diff
@@ -0,0 +1,197 @@
+all the WAL to the replica, and the replica simply ignores WAL records
+that require no action.
+
+Like in PostgreSQL, the read replica maintains a "replay LSN", which
+is the LSN up to which the replica has received and replayed the
+WAL. The replica can lag behind the primary, if it cannot quite keep
+up with the primary, or if a long-running query conflicts with changes
+that are about to be applied, or even intentionally if the user wishes
+to see delayed data (see recovery_min_apply_delay). It's important
+that the replica sees a consistent view of the whole cluster at the
+replay LSN, when it's lagging behind.
+
+In Neon, the replica connects to a safekeeper to get the WAL
+stream. That means that the safekeepers must be able to regurgitate
+the original WAL as far back as the replay LSN of any running read
+replica. (A static read-only node that does not follow the primary
+does not require a WAL stream however). The primary does not need to
+be running, and when it is, the replicas don't incur any extra
+overhead to the primary (see hot standby feedback though).
+
+### In-progress transactions
+
+In PostgreSQL, when a hot standby server starts up, it cannot
+immediately open up for queries. It first needs to establish a
+complete list of in-progress transactions, including subtransactions,
+that are running at the primary, at the current replay LSN. Normally
+that happens quickly, when the replica sees a "running-xacts" or
+shutdown checkpoint WAL record, because the primary writes a
+running-xacts record at every checkpoint, and in PostgreSQL the
+replica always starts the WAL replay from a checkpoint REDO point. If
+there are a lot of subtransactions in progress, however, the standby
+might need to wait for old transactions to complete before it can open
+up for queries.
+
+In Neon that problem is worse: a replica can start at any LSN, so
+there's no guarantee that it will see a running-xacts record any time
+soon. In particular, if the primary is not running when the replica is
+started, it might never see a running-xacts record.
+
+To make things worse, we initially missed this issue, and always
+started accepting queries at replica startup, even if it didn't have
+the transaction information. That could lead to incorrect query
+results and data corruption later. However, as we fixed that, we
+introduced a new problem compared to what we had before: previously
+the replica would always start up, but after fixing that bug, it migh
+tnot. In a superficious way, the old behavior was better (but could
+lead to serious issues later!). That made fixing that bug was very
+hard, because as we fixed it, we made things (superficially) worse for
+others.
+
+See https://github.com/neondatabase/neon/pull/7288 which fixed the
+bug, and follow-up PRs https://github.com/neondatabase/neon/pull/8323
+and https://github.com/neondatabase/neon/pull/8484 to try to claw back
+the cases that started to cause trouble as fixing it. As of this
```

Comment:
> Can you add a link to the issue for the remaining bug cases?

I now try to attach all known problems / questions to the https://github.com/neondatabase/neon/issues/6211

There is at least #8325, that wasn't started

Reply:
> > a replica might not immediately start up
> 
> This is a bit ambiguous: does it mean that it will succeed eventually, or that it might not succeed at all until the primary is running again?

If the primary is not running, and the replica is in the situation that it's waiting for a running-xacts or shutdown record to arrive, then the replica will not start up until the primary is started up again.

"Running" and "start up" terms are actually little imprecise. PostgreSQL replica startup goes through several stages:

1. First, the process is started up, and various initialization steps are performed, like initializing shared memory. If you try to connect to the server in this stage, you get an error: `ERROR: the database system is starting up`. This stage happens very quickly, no WAL is replayed yet. 

2. Then the server reads the checpoint record from the WAL and starts the WAL replay starting from the checkpoint. This works differently in Neon: we start the WAL replay at the basebackup LSN, not from a checkpoint! If you connect to the server in this state, you get an error: `ERROR: the database system is not yet accepting connections`. We proceed to the next stage, when the WAL replay sees a running-xacts record. Or in Neon, the "CLOG scanning" mechanism can allow us to move directly to next stage, with all the caveated listed in this RFC.

3. When the running-xacts information is established, the server start to accept connections.

From PostgreSQL's point of view, the server is already running in stage 2, even though it's not accepting connections yet. Our compute_ctl does not consider it as running until stage 3. If the transition from stage 2 to 3 doesn't happen fast enough, the control plane will mark the start operation as failed.

One approach would be to teach the control plane about that distinction so that we could keep the replica waiting for longer in stage 2, and also give any client connections the same `ERROR: the database system is not yet accepting connections` error that you get in standalone PostgreSQL in that state.

Reply:
I'll add this explanation to the doc